src/core/job.c

```diff
@@ -48,6 +48,28 @@ Job* job_new_raw(Unit *unit) {
         return j;
 }
 
+static uint32_t manager_get_new_job_id(Manager *m) {
```

Comment:
hmm, did such a collision actually happen? or is this paranoia?

(it's fine if this is paranoia. Just asking)

Reply:
That's paranoia. Just for safety.

Reply:
>  did such a collision actually happen?

They did. Here's what systemd crashes look like when it happens:
```
#1 0x00007fbcd9a117ab in kill () from /var/tmp/systemd-test.ZqAanQ/root/lib64/libc.so.6
#2 0x0000000000425f86 in crash (sig=6, siginfo=0x7ffcbf3f44b0, context=0x7ffcbf3f4380) at ../src/core/crash-handler.c:83
#3 0x00007fbcd9a8a59c in __pthread_kill_implementation () from /var/tmp/systemd-test.ZqAanQ/root/lib64/libc.so.6
#4 0x00007fbcd9a3dd86 in raise () from /var/tmp/systemd-test.ZqAanQ/root/lib64/libc.so.6
#5 0x00007fbcd9a10833 in abort () from /var/tmp/systemd-test.ZqAanQ/root/lib64/libc.so.6
#6 0x00007fbcda34a561 in log_assert_failed (text=0x7fbcdb074680 "counter == print_nr + 1", file=0x7fbcdb06fe43 <str+3> "src/core/manager.c", line=234, func=0x7fbcdb0745a0 <PRETTY_FUNCTION.manager_print_jobs_in_progress> "void manager_print_jobs_in_progress(Manager *)") at ../src/basic/log.c:870
#7 0x00007fbcdae3a7d1 in manager_print_jobs_in_progress (m=0x61c000000080) at ../src/core/manager.c:234
#8 0x00007fbcdae3a1fc in manager_dispatch_jobs_in_progress (source=0x612000006ac0, usec=308247013047, userdata=0x61c000000080) at ../src/core/manager.c:3007
#9 0x00007fbcda68511f in source_dispatch (s=0x612000006ac0) at ../src/libsystemd/sd-event/sd-event.c:3599
#10 0x00007fbcda683f5f in sd_event_dispatch (e=0x617000000080) at ../src/libsystemd/sd-event/sd-event.c:4175
#11 0x00007fbcda6867f2 in sd_event_run (e=0x617000000080, timeout=18446744073709551615) at ../src/libsystemd/sd-event/sd-event.c:4236
#12 0x00007fbcdae279ff in manager_loop (m=0x61c000000080) at ../src/core/manager.c:3068
#13 0x00000000004151bf in invoke_main_loop (m=0x61c000000080, saved_rlimit_nofile=0x7ffcbf3f5c80, saved_rlimit_memlock=0x7ffcbf3f5ca0, ret_retval=0x7ffcbf3f5d60, ret_shutdown_verb=0x7ffcbf3f5d20, ret_fds=0x7ffcbf3f5d90, ret_switch_root_dir=0x7ffcbf3f5ce0, ret_switch_root_init=0x7ffcbf3f5d00, ret_error_message=0x7ffcbf3f5d40) at ../src/core/main.c:1894 #14 0x000000000040947c in main (argc=5, argv=0x7ffcbf3f60d8) at ../src/core/main.c:2957
```

Reply:
@evverx How do you observe it? Is it reproducible?

Reply:
As far as I can remember it was last year and that particular crash happened in a long-running test container where among other things jobs were created by calling DBus methods (faster than usual). I think one of the jobs got stuck because I assume it had `JobTimeoutSec=infinity` (but I'm not 100% sure why exactly it got stuck) and at some point the job counter overflowed, another job got the same id and systemd crashed.

Reply:
> JobTimeoutSec=infinity

Hmm... I will consider that later.